docs/_pages/api/settings.md
---
title: Settings
layout: default
parent: API
nav_order: 1
---

All interfaces have the same internal solver settings. Note that the default settings have been tuned for 64bit floating point data types (i.e. `double`). If the solver is run with 32bit floating point data types (i.e. `float`) this can result in convergence issues. In this case, the tolerances have to be reduced.

| Argument                                         | Default Value                      | Description                                                                                                                         |
|:-------------------------------------------------|:-----------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|
| `rho_init`                                       | `1e-6`                             | Initial value for the primal proximal penalty parameter rho.                                                                        |
| `delta_init`                                     | `1e-4`                             | Initial value for the augmented lagrangian penalty parameter delta.                                                                 |
| `eps_abs`                                        | `1e-8`                             | Absolute tolerance.                                                                                                                 |
| `eps_rel`                                        | `1e-9`                             | Relative tolerance.                                                                                                                 |
| `check_duality_gap`                              | `true`                             | Check terminal criterion on duality gap.                                                                                            |
| `eps_duality_gap_abs`                            | `1e-8`                             | Absolute tolerance on duality gap.                                                                                                  |
| `eps_duality_gap_rel`                            | `1e-9`                             | Relative tolerance on duality gap.                                                                                                  |
| `infeasibility_threshold`                        | `0.9`                              | Threshold value for infeasibility detection.                                                                                        |
| `reg_lower_limit`                                | `1e-10`                            | Lower limit for regularization.                                                                                                     |
| `reg_finetune_lower_limit`                       | `1e-13`                            | Fine tune lower limit regularization.                                                                                               |
| `reg_finetune_primal_update_threshold`           | `7`                                | Threshold of number of no primal updates to transition to fine tune mode.                                                           |
| `reg_finetune_dual_update_threshold`             | `7`                                | Threshold of number of no dual updates to transition to fine tune mode.                                                             |
| `max_iter`                                       | `250`                              | Maximum number of iterations.                                                                                                       |
| `max_factor_retires`                             | `10`                               | Maximum number of factorization retires before failure.                                                                             |
| `preconditioner_scale_cost`                      | `false`                            | Scale cost in Ruiz preconditioner.                                                                                                  |
| `preconditioner_reuse_on_update`                 | `false`                            | Reuse the preconditioner from previous setup/update.                                                                                |
| `preconditioner_iter`                            | `10`                               | Maximum of preconditioner iterations.                                                                                               |
| `tau`                                            | `0.99`                             | Maximum interior point step length.                                                                                                 |
| `kkt_solver`                                     | `dense_cholesky`/<br>`sparse_ldlt` | KKT solver backend. Possible values for the dense solver: `dense_cholesky`<br>sparse solver: `sparse_ldlt`, `sparse_multistage`, `sparse_multistage_parallel` |
| `num_threads`                                    | `0`                                | Number of OpenMP threads if built with OpenMP. `0` uses the OpenMP default (e.g. `OMP_NUM_THREADS`). See [Multithreading]({{site.baseurl}}/multistage#multithreading). |
| `iterative_refinement_always_enabled`            | `false`                            | Always run iterative refinement and not only on factorization failure.                                                              |
| `iterative_refinement_eps_abs`                   | `1e-12`                            | Iterative refinement absolute tolerance.                                                                                            |
| `iterative_refinement_eps_rel`                   | `1e-12`                            | Iterative refinement relative tolerance.                                                                                            |
| `iterative_refinement_max_iter`                  | `10`                               | Maximum number of iterations for iterative refinement.                                                                              |
| `iterative_refinement_min_improvement_rate`      | `5.0`                              | Minimum improvement rate for iterative refinement.                                                                                  |
| `iterative_refinement_static_regularization_eps` | `1e-8`                             | Static regularization for KKT system for iterative refinement.                                                                      |
| `iterative_refinement_static_regularization_rel` | `eps^2`                            | Static regularization w.r.t. the maximum abs diagonal term of KKT system.                                                           |
| `verbose`                                        | `false`                            | Verbose printing.                                                                                                                   |
| `compute_timings`                                | `false`                            | Measure timing information internally.                                                                                              |


docs/interfaces/settings.md
---
title: Settings
layout: default
parent: Interfaces
nav_order: 10
---

All interfaces have the same internal solver settings. Note that the default settings have been tuned for 64bit floating point data types (i.e. `double`). If the solver is run with 32bit floating point data types (i.e. `float`) this can result in convergence issues. In this case, the tolerances have to be reduced.

| Argument                                         | Default Value                      | Description                                                                                                                         |
|:-------------------------------------------------|:-----------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|
| `rho_init`                                       | `1e-6`                             | Initial value for the primal proximal penalty parameter rho.                                                                        |
| `delta_init`                                     | `1e-4`                             | Initial value for the augmented lagrangian penalty parameter delta.                                                                 |
| `eps_abs`                                        | `1e-8`                             | Absolute tolerance.                                                                                                                 |
| `eps_rel`                                        | `1e-9`                             | Relative tolerance.                                                                                                                 |
| `check_duality_gap`                              | `true`                             | Check terminal criterion on duality gap.                                                                                            |
| `eps_duality_gap_abs`                            | `1e-8`                             | Absolute tolerance on duality gap.                                                                                                  |
| `eps_duality_gap_rel`                            | `1e-9`                             | Relative tolerance on duality gap.                                                                                                  |
| `reg_lower_limit`                                | `1e-10`                            | Lower limit for regularization.                                                                                                     |
| `reg_finetune_lower_limit`                       | `1e-13`                            | Fine tune lower limit regularization.                                                                                               |
| `reg_finetune_primal_update_threshold`           | `7`                                | Threshold of number of no primal updates to transition to fine tune mode.                                                           |
| `reg_finetune_dual_update_threshold`             | `5`                                | Threshold of number of no dual updates to transition to fine tune mode.                                                             |
| `max_iter`                                       | `250`                              | Maximum number of iterations.                                                                                                       |
| `max_factor_retires`                             | `10`                               | Maximum number of factorization retires before failure.                                                                             |
| `preconditioner_scale_cost`                      | `false`                            | Scale cost in Ruiz preconditioner.                                                                                                  |
| `preconditioner_iter`                            | `10`                               | Maximum of preconditioner iterations.                                                                                               |
| `tau`                                            | `0.99`                             | Maximum interior point step length.                                                                                                 |
| `kkt_solver`                                     | `dense_cholesky`/<br>`sparse_ldlt` | KKT solver backend. Possible values for the dense solver: `dense_cholesky`<br>sparse solver: `sparse_ldlt`, `blocksparse_stagewise` |
| `iterative_refinement_always_enabled`            | `false`                            | Always run iterative refinement and not only on factorization failure.                                                              |
| `iterative_refinement_eps_abs`                   | `1e-12`                            | Iterative refinement absolute tolerance.                                                                                            |
| `iterative_refinement_eps_rel`                   | `1e-12`                            | Iterative refinement relative tolerance.                                                                                            |
| `iterative_refinement_max_iter`                  | `10`                               | Maximum number of iterations for iterative refinement.                                                                              |
| `iterative_refinement_min_improvement_rate`      | `5.0`                              | Minimum improvement rate for iterative refinement.                                                                                  |
| `iterative_refinement_static_regularization_eps` | `1e-7`                             | Static regularization for KKT system for iterative refinement.                                                                      |
| `iterative_refinement_static_regularization_rel` | `eps^2`                            | Static regularization w.r.t. the maximum abs diagonal term of KKT system.                                                           |
| `verbose`                                        | `false`                            | Verbose printing.                                                                                                                   |
| `compute_timings`                                | `false`                            | Measure timing information internally.                                                                                              |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed
- Install `piqp.dll` and `piqpc.dll` on Windows

## [0.6.4] - 2026-08-28

### Fixed
- Fix dual recovery out-of-bounds memory bug [#42](https://github.com/PREDICT-EPFL/piqp/issues/42)

## [0.6.3] - 2026-05-11

### Changed
- Improve C++ compile times by moving large implementation details into `.tpp` files [#38](https://github.com/PREDICT-EPFL/piqp/pull/38)

### Fixed
- Fix compatibility with Eigen 5 [#41](https://github.com/PREDICT-EPFL/piqp/pull/41)
- Publish Octave release tarballs with versioned filenames [#36](https://github.com/PREDICT-EPFL/piqp/issues/36)

## [0.6.2] - 2025-09-16

### Changed
- Record KKT timings
- Disable stack execution for matlab and python libraries on linux [#35](https://github.com/PREDICT-EPFL/piqp/pull/35)

## [0.6.1] - 2025-08-22

### Changed
- Add tracy support [#30](https://github.com/PREDICT-EPFL/piqp/pull/30)
- Enforce Eigen ABI with Template Instantiation [#32](https://github.com/PREDICT-EPFL/piqp/pull/32)
- Improved and more robust infeasibility detection [#34](https://github.com/PREDICT-EPFL/piqp/pull/34)

## [0.6.0] - 2025-06-30

### Changed
- Fix compatibility with Octave 10 or newer [#20](https://github.com/PREDICT-EPFL/piqp/pull/20)
- Change preconditioner to scale lower and upper bounds using the same scaling and refactor KKT system logic [#21](https://github.com/PREDICT-EPFL/piqp/pull/21)
- Add support for double sided general inequality constraints [#22](https://github.com/PREDICT-EPFL/piqp/pull/22)
- General convergence improvements [#23](https://github.com/PREDICT-EPFL/piqp/pull/23)
- Switch build system to scikit-build-core [#25](https://github.com/PREDICT-EPFL/piqp/pull/25)
- Make building with EIGEN_MAX_ALIGN_BYTES definition opt-in [#28](https://github.com/PREDICT-EPFL/piqp/pull/28)

## [0.5.0] - 2025-03-17

### Changed
- In the sparse interface for Python A, b, G, and h are now None by default
- Use Cholesky decomposition instead of LDLt factorization in dense interface
- New KKT solver backend `sparse_multistage` for multistage optimization problems [#14](https://github.com/PREDICT-EPFL/piqp/pull/14)
- Fix with run_time calculation, i.e., now it's run_time = setup_time/update_time + solve_time and not accumulated over multiple runs.

## [0.4.2] - 2024-08-02

### Changed

- Handle inf constraints in general inequalities by setting corresponding row in G to zero and emitting a warning to the user.

## [0.4.1] - 2024-06-22

### Changed

- Fix installation paths for CMake install.

## [0.4.0] - 2024-06-21

### Changed

- Instead of building both shared and static libraries, it can now be selected by setting `BUILD_SHARED_LIBS`.
- Better support for CMake subdirectory installs.
- Added Octave interface (thanks to @redstone99)
- Added boundary control, i.e., dual inequality variables can't become zero anymore avoiding potential NaNs in the solution.

## [0.3.1] - 2024-05-25

### Changed

- Added python stub files for better python IDE support.

## [0.3.0] - 2024-05-24

### Changed

- Make equality and inequality constrains optional on setup.
- Pre-compile common template instantiations in C++ interface to speed up compilation times.
- Add utilities to save and load problem data.
- Various bug fixes.
- Various doc improvements.

## [0.2.4] - 2023-12-25

### Changed

- Fixed issue where regularization parameters were not correctly reset after an update.

## [0.2.3] - 2023-09-14

### Changed

- Allow compilation with custom scalar types like mpreal.
- Disable floating point contractions in sparse ldlt.

## [0.2.2] - 2023-08-04

### Changed

- Fixed compilation issues for conda-forge on macOS

## [0.2.1] - 2023-08-03

### Changed

- Renamed C libraries from `piqp` to `piqpc` and `piqpcstatic` to avoid naming conflicts (especially on Windows).
- Pull external packages through FetchContent instead of git submodules.

## [0.2.0] - 2023-07-28

### Changed

- Added Matlab interface.
- The backend is now printed in verbose mode.

## [0.1.3] - 2023-07-20

### Changed

- Changed print function for potential Matlab and R interfaces.

## [0.1.2] - 2023-07-14

### Changed

- Fixed PyPi upload.
- Fixed incorrect type in settings.
- Added runtime dimension checks in release mode.
- Convert infinite inequality bounds to a finite value internally (-inf -> -1e30, inf -> 1e30).

## [0.1.1] - 2023-06-30

### Changed

- Fixed compilation issues on Windows.
- Removed French accents in verbose mode prints which caused issues in certain consoles, i.e. Windows.
- Exposed fine tune thresholds in settings.
- Fine-tuned default settings for better numerical robustness.

## [0.1.0] - 2023-06-28

Initial release of PIQP!

[unreleased]: https://github.com/PREDICT-EPFL/piqp/compare/v0.6.4...HEAD
[0.6.4]: https://github.com/PREDICT-EPFL/piqp/compare/v0.6.2...v0.6.4
[0.6.3]: https://github.com/PREDICT-EPFL/piqp/compare/v0.6.2...v0.6.3
[0.6.2]: https://github.com/PREDICT-EPFL/piqp/compare/v0.6.1...v0.6.2
[0.6.1]: https://github.com/PREDICT-EPFL/piqp/compare/v0.6.0...v0.6.1
[0.6.0]: https://github.com/PREDICT-EPFL/piqp/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/PREDICT-EPFL/piqp/compare/v0.4.2...v0.5.0
[0.4.2]: https://github.com/PREDICT-EPFL/piqp/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/PREDICT-EPFL/piqp/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/PREDICT-EPFL/piqp/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com/PREDICT-EPFL/piqp/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/PREDICT-EPFL/piqp/compare/v0.2.4...v0.3.0
[0.2.4]: https://github.com/PREDICT-EPFL/piqp/compare/v0.2.3...v0.2.4
[0.2.3]: https://github.com/PREDICT-EPFL/piqp/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/PREDICT-EPFL/piqp/compare/v0.2.1...v0.2.2

include/piqp/settings.hpp
// This file is part of PIQP.
//
// Copyright (c) 2024 EPFL
// Copyright (c) 2022 INRIA
//
// This source code is licensed under the BSD 2-Clause License found in the
// LICENSE file in the root directory of this source tree.

#ifndef PIQP_SETTINGS_HPP
#define PIQP_SETTINGS_HPP

#include <limits>
#include "piqp/typedefs.hpp"

namespace piqp
{

enum class KKTSolver
{
    dense_cholesky,
    sparse_ldlt,
    sparse_ldlt_eq_cond,
    sparse_ldlt_ineq_cond,
    sparse_ldlt_cond,
    sparse_multistage,
    sparse_multistage_parallel
};

constexpr const char* kkt_solver_to_string(KKTSolver kkt_solver)
{
    switch (kkt_solver)
    {
        case KKTSolver::dense_cholesky: return "dense_cholesky";
        case KKTSolver::sparse_ldlt: return "sparse_ldlt";
        case KKTSolver::sparse_ldlt_eq_cond: return "sparse_ldlt_eq_cond";
        case KKTSolver::sparse_ldlt_ineq_cond: return "sparse_ldlt_ineq_cond";
        case KKTSolver::sparse_ldlt_cond: return "sparse_ldlt_cond";
        case KKTSolver::sparse_multistage: return "sparse_multistage";
        case KKTSolver::sparse_multistage_parallel: return "sparse_multistage_parallel";
        default: return "unknown";
    }
}

template<typename T>
struct Settings
{
    T rho_init = 1e-6;
    T delta_init = 1e-4;

    T eps_abs = 1e-8;
    T eps_rel = 1e-9;

    bool check_duality_gap = true;
    T eps_duality_gap_abs = 1e-8;
    T eps_duality_gap_rel = 1e-9;

    T infeasibility_threshold = 0.9;

    T reg_lower_limit = 1e-10;
    T reg_finetune_lower_limit = 1e-13;
    isize reg_finetune_primal_update_threshold = 7;
    isize reg_finetune_dual_update_threshold = 7;

    isize max_iter = 250;
    isize max_factor_retires = 10;

    bool preconditioner_scale_cost = false;
    bool preconditioner_reuse_on_update = false;
    isize preconditioner_iter = 10;

    T tau = 0.99;

    KKTSolver kkt_solver = KKTSolver::dense_cholesky;

    isize num_threads = 0;

    bool iterative_refinement_always_enabled = false;
    T iterative_refinement_eps_abs = 1e-12;
    T iterative_refinement_eps_rel = 1e-12;
    isize iterative_refinement_max_iter = 10;
    T iterative_refinement_min_improvement_rate = 5.0;
    T iterative_refinement_static_regularization_eps = 1e-8;
    T iterative_refinement_static_regularization_rel = std::numeric_limits<T>::epsilon() * std::numeric_limits<T>::epsilon();

    bool verbose = false;
    bool compute_timings = false;

    bool verify_settings() const noexcept
    {
        return rho_init > 0 &&
               delta_init > 0 &&
               eps_abs > 0 &&
               eps_rel >= 0 &&
               eps_duality_gap_abs > 0 &&
               eps_duality_gap_rel >= 0 &&
               infeasibility_threshold >= 0 &&
               reg_lower_limit > 0 &&
               reg_finetune_primal_update_threshold >= 0 &&
               reg_finetune_dual_update_threshold >= 0 &&
               max_iter > 0 &&
               max_factor_retires > 0 &&
               preconditioner_iter >= 0 &&
               tau > 0 && tau <= 1 &&
               num_threads >= 0 &&
               iterative_refinement_eps_abs > 0 &&
               iterative_refinement_eps_rel >= 0 &&
               iterative_refinement_max_iter >= 0 &&
               iterative_refinement_min_improvement_rate >= 1.0 &&
               iterative_refinement_static_regularization_eps > 0 &&
               iterative_refinement_static_regularization_rel >= 0;
    }
};

} // namespace piqp

#endif //PIQP_SETTINGS_HPP
